### DISCUSSION TITLE
Minimal Linux Implementation Using Rust Coreutils

### DISCUSSION THREAD

**Discussion by Raghav-45** [uutils/coreutils#6457] [Q&A]
Hey,

I'm working on a project named Wave Linux. You can take a look [here](https://github.com/Raghav-45/wave-linux). When I was in 10th grade (16 years old), I made a minimal Linux. In that project, I used the BusyBox multicall binary, which I extended by creating symlinks for commands like `ls`, `rm`, and `echo`.

Currently, I'm 19 and in college, and I'm working on my current hobby project - Wave Linux. For this new project, I want to start my journey with your Rust Coreutils (not the GNU Coreutils). However, when I build it using the provided steps, it only creates a single executable file( **.exe** ). I don't think this will work like the BusyBox multicall binary.

Could you please guide me on how to achieve a BusyBox-like multicall binary using Rust Coreutils?

I would love and greatly appreciate it if anyone would like to contribute or help me build this Linux. I'm also planning to add graphical support so that we can move beyond the command line interface. The goal of this project is to create a protected operating system similar to macOS, but with features like Windows.

Thanks!

**Comment by BenWiederhake** [comment]
Can you clarify what you mean? On Linux, `cargo build` already builds a multicall binary by default.

**Comment by Raghav-45** [comment]
> `cargo build` already builds a multicall binary by default.

Yeah, It does build a multicall binary, it produces an `.exe` file. But for building a minimal Linux, I need a multiplatform binary file that can run on various architectures. Like busybox static binary which we used for. Designing minimal Linux using `musl-gcc`

![Screenshot_2024-06-10-12-08-54-15_f9ee0578fe1cc94de7482bd41accb329~3.jpg](https://github.com/uutils/coreutils/assets/77260113/fae186bf-bddf-4332-9bd0-770cdfa6fc20)

And after making busybox binary, I used to copy it in this directory:
![IMG_20240610_122246.jpg](https://github.com/uutils/coreutils/assets/77260113/c5dbe6b2-d6fd-4432-a16a-e7ccc540e6c2)


Could you provide guidance on how to # achieve this with Rust Coreutils?

**Reply by tertsdiepraam** [reply]
Is this what you're looking for? https://doc.rust-lang.org/reference/linkage.html#static-and-dynamic-c-runtimes

You could compile with musl and that could give you a static lib probably. Although it might not work with all utils.

**Reply by Raghav-45** [reply]
> Is this what you're looking for? https://doc.rust-lang.org/reference/linkage.html#static-and-dynamic-c-runtimes

I've visited this link and really thanks for your efforts, I think I got my answer, I was talking about cross-compiling these utils


> 
> You could compile with musl and that could give you a static lib probably. Although it might not work with all utils.

Yes, I'm talking about cross-compiling. But Is there an easy way to do this?

Coz I have used **musl-gcc** to cross-compile BusyBox before, which gave me a static BusyBox binary.

That standalone busybox binary provided utility commands in my minimal Linux, such as `busybox ls` and `busybox cd /path`.

Do you know any better approach of doing these?

Edit: this is what I found online, I'll inform you if it works for me:
`cargo build --target x86_64-unknown-linux-musl --release`

**Reply by tertsdiepraam** [reply]
Yes that's the way to do it!

**Reply by Raghav-45** [reply]
Hey, I have some exciting things to share, I've got it running as a **single standalone binary** for coreutils. Now we can say _Good Bye_ to GNU Coreutils and their licensing standards.

see:
![Screenshot (236)](https://github.com/uutils/coreutils/assets/77260113/59534d6e-8c2e-495a-8550-dd0988e66218)

And this is how it is working:
![Screenshot (237)](https://github.com/uutils/coreutils/assets/77260113/cb062c47-310a-4b16-a4fd-6019f9c00c11)

thanks for making this awsome project

For those who want to know how I built this single standalone binary for minimal Linux, follow these steps:
```shell
rustup target add x86_64-unknown-linux-musl
```

```shell
cargo build --target x86_64-unknown-linux-musl --release
```

If you encounter the error: `error occurred: Failed to find tool. Is **musl-gcc** installed?`

Just install `musl-tools`

Again Thanks for making this awsome project
